### ISSUE TITLE
New replicas which error while creating data on disk should be unregistered

### ISSUE THREAD

**Issue by alanking** [irods/irods#4958]
- [ ] master
- [ ] 4-2-stable
---
## Bug Report

When a replica returns an error while being created on disk, rsDataObjCreate unlinks the file and unregisters the replica from the catalog:
https://github.com/irods/irods/blob/847574a932bd1e73a04745447b462809dfb5eee1/server/api/src/rsDataObjCreate.cpp#L57-L64

### What did you try to do?
```
$ ilsresc -l nowrite
resource name: nowrite
id: 18435
zone: tempZone
type: unixfilesystem
class: cache
location: icat.example.org
vault: /var/lib/irods/nowritevault
free space: 
free space time: : Never
status: 
info: 
comment: 
create time: 01591029793: 2020-06-01.16:43:13
modify time: 01591029793: 2020-06-01.16:43:13
context: 
parent: 
parent context: 
$ chmod 644 nowritevault
$ ls -al nowritevault/
ls: cannot access 'nowritevault/..': Permission denied
ls: cannot access 'nowritevault/.': Permission denied
total 0
d????????? ? ? ? ?            ? .
d????????? ? ? ? ?            ? ..
```

### Expected behavior
The replica should remain in the catalog regardless of the state of the data on disk. The replica status should be updated to reflect its stale state and the client can decide how to proceed (i.e. retry or give up).
```
$ iput -R nowrite file nowrite_file
remote addresses: 192.168.96.2 ERROR: putUtil: put error for /tempZone/home/rods/nowrite_file, status = -526013 status = -526013 UNIX_FILE_GET_FS_FREESPACE_ERR, Permission denied
$ ils -l nowrite_file
  rods              0 nowrite            0 2020-06-01.16:43 X nowrite_file
```

### Observed behavior (including steps to reproduce, if applicable)
```
$ iput -R nowrite file nowrite_file
remote addresses: 192.168.96.2 ERROR: putUtil: put error for /tempZone/home/rods/nowrite_file, status = -526013 status = -526013 UNIX_FILE_GET_FS_FREESPACE_ERR, Permission denied
$ ils -l nowrite_file
remote addresses: 192.168.96.2 ERROR: lsUtil: srcPath /tempZone/home/rods/nowrite_file does not exist or user lacks access permission
```

**Comment by trel** [comment]
This has appeared in CU Boulder's deployment when using NFS as a storage backend.   If the NFS mountpoint goes stale, this error path becomes important (and noisy).

**Comment by peggy0135** [comment]
Is there any way that we can solve it.
Did it happen with ceph block device?
Thank you.

**Comment by trel** [comment]
We will solve it by probably taking the approach noted above as 'Expected behavior'.

In the meantime, no, this is the behavior - if the create fails with an error, you'll see the unlink and unregister attempts probably fail in the logs.

I don't know whether this was a ceph block device.  I would guess not, but am not sure.

**Comment by alanking** [comment]
I was not using a Ceph block device at the time of making this issue.

**Comment by alanking** [comment]
Upon re-re-examination, it seems better to cleanup on a broken create.

**Comment by alanking** [comment]
This was fixed in PRs #5539 and #5547

The relevant commits are...
master: https://github.com/irods/irods/commit/1c54a98fcd90a69ef49ab8333a6002ab20d15137
4-2-stable: https://github.com/irods/irods/commit/f65ce76c436948c91a5e7eac2dfceed5f38f55ca

Relevant lines (master and 4-2-stable are aligned here save for line numbers):
https://github.com/irods/irods/blob/f65ce76c436948c91a5e7eac2dfceed5f38f55ca/server/api/src/rsDataObjOpen.cpp#L380-L400

Checkboxing and closing.
